fix: omit malicious address alerts on revoke

imblue-dabadee opened the pull request:
<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been completely filled out, and PR status checks have passed at least once.
-->

## **Description**

<!--
Write a short description of the changes included in this pull request, also include relevant motivation and context. Have in mind the following questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

Currently, we make the user acknowledge a malicious address via an alert when a user is revoking an on-chain permission for a malicious address. This creates unnecessary friction for users performing the safe action of removing a permission.

This PR suppresses the blocking spender trust signal alerts when the transaction or signature is detected as a revoke operation. The inline address label is still displayed as informational, but users no longer need to confirm the alert.

Revoke detection covers:
- `setApprovalForAll` with `approved: false`
- ERC20 `approve` with zero amount
- Permit signatures with `value: '0'`
- DAI permit with `allowed: false`

Have also handled NFT approvals with `tokenId: 0` so that they are not incorrectly treated as ERC20 revokes and we still alert the user that they giving approval to a malicious address.

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/38992?quickstart=1)

## **Changelog**

<!--
If this PR is not End-User-Facing and should not show up in the CHANGELOG, you can choose to either:
1. Write `CHANGELOG entry: null`
2. Label with `no-changelog`

If this PR is End-User-Facing, please write a short User-Facing description in the past tense like:
`CHANGELOG entry: Added a new tab for users to see their NFTs`
`CHANGELOG entry: Fixed a bug that was causing some NFTs to flicker`

(This helps the Release Engineer do their job more quickly and accurately)
-->

CHANGELOG entry: Fixed malicious address alert incorrectly blocking revoke transactions

## **Related issues**

Fixes: https://consensyssoftware.atlassian.net/browse/PSAFE-182

## **Manual testing steps**

1. Message myself for a kitchen sink to test against.

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

#### ERC20: `approve(spender,0)`
<img width="395" height="622" alt="Screenshot 2025-12-19 at 4 49 19 PM" src="https://github.com/user-attachments/assets/267e1ff4-d76e-4e3f-9378-e80d5500b731" />

#### ERC20: `increaseAllowance(0)`
<img width="407" height="615" alt="Screenshot 2025-12-19 at 4 49 54 PM" src="https://github.com/user-attachments/assets/7e01ec96-309b-46c0-bc35-ae0c67bcb1fb" />

#### ERC721: `setApprovalForAll(op, false)`
<img width="402" height="634" alt="Screenshot 2025-12-19 at 4 54 38 PM" src="https://github.com/user-attachments/assets/991c43db-5ab3-43d2-8367-89173a5dcbea" />

#### ERC721: Approving a malicious address for when the token ID is 0
<img width="394" height="626" alt="Screenshot 2025-12-19 at 4 41 52 PM" src="https://github.com/user-attachments/assets/8a43cf95-0074-4506-8f6b-955b9c7945df" />

#### DAI Permit with `allowed: false`
<img width="401" height="621" alt="Screenshot 2026-01-20 at 12 09 43 PM" src="https://github.com/user-attachments/assets/155a98b0-148d-4d8f-aee8-7378371fadd7" />

#### EIP-2612 Permit signature with value: '0'
<img width="399" height="616" alt="Screenshot 2026-01-20 at 12 11 32 PM" src="https://github.com/user-attachments/assets/cb746881-19ef-461e-8aa5-28afdac79c67" />


### **After**

<!-- [screenshots/recordings] -->

#### ERC20: `approve(spender,0)`
<img width="398" height="618" alt="Screenshot 2025-12-19 at 4 30 30 PM" src="https://github.com/user-attachments/assets/e0e4f9d8-eb95-4d60-850d-a10f73795b06" />

#### ERC20: `increaseAllowance(0)`
<img width="397" height="589" alt="Screenshot 2025-12-19 at 4 33 48 PM" src="https://github.com/user-attachments/assets/a9800fe6-91a5-4a28-8939-443d8507b6b2" />

#### ERC721: `setApprovalForAll(op, false)`
<img width="400" height="624" alt="Screenshot 2025-12-19 at 4 33 02 PM" src="https://github.com/user-attachments/assets/6036a04c-09e7-4475-8055-647eaf593d25" />

#### ERC721: Approving a malicious address for when the token ID is 0
<img width="394" height="626" alt="Screenshot 2025-12-19 at 4 41 52 PM" src="https://github.com/user-attachments/assets/8a43cf95-0074-4506-8f6b-955b9c7945df" />

#### DAI Permit with `allowed: false`
<img width="400" height="621" alt="Screenshot 2026-01-20 at 12 10 57 PM" src="https://github.com/user-attachments/assets/3a95f8ff-8e4d-437f-92a7-9e8250afd2b1" />

#### EIP-2612 Permit signature with value: '0'
<img width="399" height="615" alt="Screenshot 2026-01-20 at 12 12 49 PM" src="https://github.com/user-attachments/assets/6fb82ca0-4f7d-4766-8025-dc51aeac54a6" />


## **Pre-merge author checklist**

- [x] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.


<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Reduces false-positive/blocking alerts for safe revoke flows while preserving alerts for risky approvals.
> 
> - Implements revoke detection in `useSpenderAlerts` for `approve(0)`, `setApprovalForAll(false)`, EIP-2612 `value: '0'`, and DAI `allowed: false`
> - Adds NFT-aware logic so zero tokenId NFT approvals still alert (uses `useIsNFT` and token standard lookup via `getTokenStandardAndDetailsByChain` with `useAsyncResult`); supports Permit2 flows
> - Returns alerts only for malicious/warning trust signals when not a revoke; skips when zero-value unless NFT or explicitly zero-value
> - Extensive new tests for `useSpenderAlerts` (ERC20/ERC721/Permit/Permit2, pending states); test suites updated to mock async hooks and avoid act warnings
> - Snapshot updates: replaces skeleton headers with static titles/subtitles (e.g., "Spending cap request", "Signature request"); minor numeric formatting change (`30` → `3,000`)
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit f64013557b84ea1dd2e826c35583339a1da90b89. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

imblue-dabadee committed:
feat: skip malicious address alert on revoke operations for approvals and permits

imblue-dabadee committed:
feat: skip malicious address alert friction when revoking permissions

imblue-dabadee committed:
chore: use object destructuring

imblue-dabadee committed:
test: conform to guidelines

imblue-dabadee committed:
Revert "test: conform to guidelines"

This reverts commit 9d50fe56b1c02cb18bc661dfad624fc8bf0a4c70.

imblue-dabadee committed:
update unit test names

imblue-dabadee committed:
test: improve the visibility of the exercise phase

imblue-dabadee committed:
test: clearer names

imblue-dabadee committed:
test: use consts for realistic data + test case when token is not in cache

imblue-dabadee committed:
refactor: avoid typecasting as any

imblue-dabadee committed:
Merge branch 'main' into fix/omit-address-warnings-on-revoke

imblue-dabadee committed:
refactor: useIsNft

imblue-dabadee committed:
refactor: improve readability

imblue-dabadee committed:
test: update test

imblue-dabadee committed:
refactor: change Loading enum values from 'loading' to 'Loading'

imblue-dabadee committed:
test: update fixtures to use 'Loading' enum value

imblue-dabadee committed:
refactor: use ResultType enum instead of string literals

imblue-dabadee committed:
Revert "refactor: use ResultType enum instead of string literals"

This reverts commit dfb287c0735e7a49499b62e110165c0a9901a8e4.

imblue-dabadee committed:
Revert "refactor: change Loading enum values from 'loading' to 'Loading'"

This reverts commit 9784f2e0ab1884e1d8666016205e5dec11e5d877.

imblue-dabadee committed:
Revert "test: update fixtures to use 'Loading' enum value"

This reverts commit f43352bfa61e0df1ac985c08bb7439183a48c859.

imblue-dabadee committed:
Merge branch 'main' into fix/omit-address-warnings-on-revoke

AugmentedMode reviewed (dismissed):
LGTM besides the one comment

imblue-dabadee committed:
fix: lowercase the dai CA check

imblue-dabadee committed:
fix: add case-insensitive comparison in getIsRevokeDAIPermit

imblue-dabadee committed:
feat: skip spender alerts for zero-value operations

- Add Permit2 zero-amount detection (PermitSingle, PermitBatch, PermitTransferFrom)
- Handle Permit2 approve transactions by checking token address from calldata
- Reuse existing getIsRevokeDAIPermit utility
- Rename revoke-related variables to isSafeToSkipAlert/AlertSkipReason for clarity

imblue-dabadee committed:
test: add Permit2 zero-value operation tests for useSpenderAlerts

imblue-dabadee committed:
refactor: remove Permit2 zero-amount detection from spender alerts

Permit2 signatures with amount=0 are not true revocations - they just
authorize zero token transfers. Showing alerts for these can help users
feel protected, and the scenario is unlikely in practice.

imblue-dabadee committed:
test: remove Permit2 zero-amount tests from useSpenderAlerts

imblue-dabadee committed:
fix: prevent DAI permit alert bypass via fake value field

DAI permits use the 'allowed' boolean field, not 'value'. The previous
OR logic allowed an attacker to craft a DAI permit with allowed:true
(granting permission) and a fake value:'0' field to bypass the alert.

Now we correctly check 'allowed === false' for DAI permits and
'value === 0' only for standard EIP-2612 permits.

imblue-dabadee committed:
refactor: flatten getAlertSkipReason with early returns

imblue-dabadee committed:
fix: add missing dependencies to useAsyncResult

imblue-dabadee committed:
test: refactor useSpenderAlerts tests to follow unit testing guidelines

- Remove 'should' from test names
- Replace beforeEach with setupDefaultMocks helper function
- Add helper functions to reduce code duplication
- Use simple constants for expected alerts instead of factory function

imblue-dabadee committed:
style: fix prettier formatting in useSpenderAlerts.test.ts

imblue-dabadee committed:
Merge branch 'main' into fix/omit-address-warnings-on-revoke

imblue-dabadee committed:
revert: remove case-insensitive comparison in getIsRevokeDAIPermit

imblue-dabadee committed:
test: mock async hooks to prevent React Act warnings

Add mocks for useIsNFT and useAsyncResult in confirm.test.tsx
and useConfirmationAlerts.test.ts to prevent async state updates
from triggering React Act warnings during tests.

imblue-dabadee committed:
test: add useAsyncResultOrThrow mock and update snapshots

imblue-dabadee committed:
test: add useAsyncResultOrThrow mock to useConfirmationAlerts.test.ts

imblue-dabadee committed:
test: use jest.requireActual to preserve unmocked exports

AugmentedMode reviewed (approved):
LGTM
